src/main/scala/vectorpipe/vectortile/Pipeline.scala

```diff
@@ -0,0 +1,126 @@
+ *
+ * Each of these operations is documented below.  The only extra note is that [[
+ * reduce]] and [[simplify]] will be called prior to processing the initial zoom
+ * level.
+ */
+trait Pipeline {
+  /**
+   * The root URI for output.
+   */
+  val baseOutputURI: java.net.URI
+
+  /**
+   * Name of the column containing the JTS geometry.
+   */
+  val geometryColumn: String
+
+  /**
+   * Reduce the input data between zoom levels.
+   *
+   * Not all data is useful in all levels of the vector tile pyramid.  When a
+   * geometry will no longer be used in any subsequent generated level, it may be
+   * desirable to drop the relevant rows from the DataFrame to limit the amount
+   * of working data.  Moreover, it may be desirable to collect multiple entities
+   * into a single, aggregated entity.  These sorts of operations are made
+   * possible by the [[reduce]] function.
+   *
+   * This function will be called once before tile generation begins at the
+   * initial zoom level.  All participating geometries will already have been
+   * keyed to the given layout, with the list of relevant keys available in the
+   * "keys" field.
+   *
+   * @param   input           A DataFrame minimally containing a "geom" field of
+   *                          JTS [[Geometry]] and a field of Array[SpatialKey]
+   *                          with the name given by the keyColumn param, below
+   * @param   layoutLevel     The layout level for the target zoom
+   * @param   keyColumn       The name of the column containing
+   *                          Array[SpatialKey] giving the list of keys that the
+   *                          geometry interacts with
+   */
+  def reduce(input: DataFrame, layoutLevel: LayoutLevel, keyColumn: String): DataFrame = input
+
+  /*
+   * Lower complexity of geometry while moving to less resolute zoom levels.
+   *
+   * While moving from finer to coarser levels of the pyramid, it may not be
+   * necessary to maintain the full level of detail of available geometries.
+   * This function is used to reduce the complexity of geometries.  The
+   * [[LayoutDefinition]] will be for the target zoom level.
+   *
+   * By default, we return the input geometry without simplification, but there
+   * is a simplifier using JTS's topology-preserving simplifier available in
+   * [[vectorpipe.vectortile.Simplify]].
+   */
+  def simplify(g: jts.Geometry, layout: LayoutDefinition): jts.Geometry = g
+
+  /**
+   * Select geometries for display at a given zoom level.
+   *
+   * VectorPipe allows geometries to be carried up the pyramid for display at a
+   * later time.  This function is used to choose the elements that will be
+   * displayed at the target zoom level.  This is useful for implementing a
+   * display schema, or for example, selection based on prominence (relying on
+   * the array of spatial keys added to the data frame in the column with the
+   * longest name fitting the regular expression '[_]+keys').
```

Comment:
Can you clarify the parenthetical and/or provide an example? I'm having trouble following the key-related metavariables.